framework/src/actions/CreateProblemAction.C

```diff
@@ -20,131 +18,34 @@ template <>
-      if (_app.useEigenvalue() && _type != "EigenProblem" &&
-          !(std::dynamic_pointer_cast<EigenProblem>(_problem)))
-        mooseError("Problem has to be of a EigenProblem (or derived subclass) type when using "
-                   "eigen executioner");
-    }
-    // set up the problem
-    _problem->setCoordSystem(_blocks, _coord_sys);
-    _problem->setAxisymmetricCoordAxis(getParam<MooseEnum>("rz_coord_axis"));
-    _problem->setKernelCoverageCheck(getParam<bool>("kernel_coverage_check"));
-    _problem->setMaterialCoverageCheck(getParam<bool>("material_coverage_check"));
-    _problem->setParallelBarrierMessaging(getParam<bool>("parallel_barrier_messaging"));
-
-    if (isParamValid("restart_file_base"))
-    {
-      std::string restart_file_base = getParam<FileNameNoExtension>("restart_file_base");
-
-      std::size_t slash_pos = restart_file_base.find_last_of("/");
-      std::string path = restart_file_base.substr(0, slash_pos);
-      std::string file = restart_file_base.substr(slash_pos + 1);
-
-      /**
-       * If the user specified LATEST as the file in their directory path, find the file with the
-       * latest timestep and the largest serial number.
-       */
-      if (file == "LATEST")
-      {
-        std::list<std::string> dir_list(1, path);
-        std::list<std::string> files = MooseUtils::getFilesInDirs(dir_list);
-        restart_file_base = MooseUtils::getLatestAppCheckpointFileBase(files);
+    // when this action is built by parser with Problem input block, this action
+    // must act i.e. create a problem. Thus if a problem has been created, it will error out.
+    if (_problem)
+      mooseError("Trying to build a problem but problem has already existed");
 
-        if (restart_file_base == "")
-          mooseError("Unable to find suitable restart file");
-      }
+    _moose_object_pars.set<MooseMesh *>("mesh") = _mesh.get();
+    _moose_object_pars.set<bool>("use_nonlinear") = _app.useNonlinear();
 
-      _console << "\nUsing " << restart_file_base << " for restart.\n\n";
-      _problem->setRestartFile(restart_file_base);
-    }
+    _problem =
+        _factory.create<FEProblemBase>(_type, getParam<std::string>("name"), _moose_object_pars);
 
-    // Create etra vectors and matrices if any
-    CreateTagVectors();
-  }
-}
-
-void
-CreateProblemAction::CreateTagVectors()
-{
-  // add vectors and their tags to system
-  auto & vectors = getParam<std::vector<TagName>>("extra_tag_vectors");
-  auto & nl = _problem->getNonlinearSystemBase();
-  for (auto & vector : vectors)
-  {
-    auto tag = _problem->addVectorTag(vector);
-    nl.addVector(tag, false, GHOSTED);
-  }
-
-  // add matrices and their tags
-  auto & matrices = getParam<std::vector<TagName>>("extra_tag_matrices");
-  for (auto & matrix : matrices)
-  {
-    auto tag = _problem->addMatrixTag(matrix);
-    nl.addMatrix(tag);
+    if (!_problem.get())
+      mooseError("Problem has to be of a FEProblemBase type");
```

Comment:
May be I am missing something here, but if we do `_problem = _factory.create<FEProblemBase>(...)`, how can `_problem.get()` be `nullptr`? If `Factory` is building something that is not `FEProblemBase`-derived, it bombs out inside `Factory::create`, right?

Reply:
Good question. I do not know the answer cause i copy-pasted this. @permcody any thoughts? I think here we want to make sure users are giving a valid `type`, otherwise MOOSE will throw out a meaningful error message.

Reply:
I think this check is not necessary, I will remove it.